app/controllers/orders_controller.rb

```diff
@@ -0,0 +1,233 @@
+class OrdersController < ApplicationController
+  def new
+    authorize Order
+    @order = Order.new
+  end
+
+  def checkout
+    authorize Order
+    @order = Order.new
+    @cart = Cart.find_by(user_id: current_user.id)
+    @total_amount, @total_discount, @discounts = calculate_total_amount
+  end
+
+  def create
+    authorize Order
+    @order = Order.new
+    # we have to explicitly read and save every parameter because the expiry month/year are coming in as
+    # separate parameters but should be unified in one field before saving to the database
+    @order.user_id = order_params[:user_id]
+    @order.credit_card_type = order_params[:credit_card_type]
+    @order.credit_card_name = order_params[:credit_card_name]
+    @order.credit_card_number = order_params[:credit_card_number].gsub(/\s+/, "")
+    @order.credit_card_cvc = order_params[:credit_card_cvc]
+    @order.credit_card_expiry = Date.new(
+        order_params[:credit_card_expiry_year].to_i,
+        order_params[:credit_card_expiry_month].to_i, 1
+    )
+
+    # get the contents of the cart
+    @cart = Cart.find(current_user.cart_id)
+
+    # for security reasons, calculate the total amount from the actual cart (instead of taking a parameter)
+    @total_amount, @total_discount, @discounts = calculate_total_amount
+    @order.total_amount = @total_amount
+
+    logger.debug "Total amount is #{@total_amount}"
+    logger.debug "Trying to create new product with order #{@order.inspect}"
+
+    # before creating anything, check if there is enough stock of all requested products
+    # in case there is a problematic item, it will be saved to @cart_item
+    @cart_item = nil
+    @cart.cart_items.each do |cart_item|
+      if cart_item.product.stock_level < cart_item.amount
+        @cart_item = cart_item
+      end
+    end
+
+    # save order object without running validations yet, so we get the object ID
+    # validations will be run below
+    @order.save(:validate => false)
+
+    # create order_items out of all cart_items
+    # decrease stock level of all cart_items by amount ordered
+    # then remove order items
+    @cart.cart_items.each do |cart_item|
+      cart_item.becomes!(OrderItem)
+      cart_item.update!(order_id: @order.id, cart_id: nil)
+      cart_item.product.stock_level -= 1
+      cart_item.product.save!
+    end
+
+    respond_to do |format|
+      if @order.save and @cart_item.nil?
+        format.html { redirect_to @order, notice: 'Your order has been placed.' }
+        format.json { render :show, status: :created, location: @order }
+      else
+        # if save is unsuccessful, roll back changes from above
+        @order.order_items.each do |order_item|
+          order_item.becomes!(CartItem)
+          order_item.update!(cart_id: @cart.id, order_id: nil)
+          order_item.product.stock_level += 1
+          order_item.product.save!
+        end
+        @order.destroy
```

Comment:
Q24

Consider using Rail's Transaction support, so the rollback can be done for you. See for example http://api.rubyonrails.org/classes/ActiveRecord/Rollback.html